[sharding] Extract grpc feature into sub-crate

agourlay opened the pull request:
This PR extracts the grpc types into a sub-crate in order to enable remote shards to use the grpc API #407 

## Context

Prior to this change, the crate dependency tree looks like this:

```mermaid
graph TD;
      segment-->storage;
      collection-->storage;
      segment-->collection;
      segment --> root;
      storage --> root;
      collection --> root;
```

The `segment` crate depends on nothing and the `root` crate composes all crates to build the REST & grpc APIs.

The goal of this PR is to enable the `RemoteShard` in the `collection` crate to perform grpc calls.

## Solution

Extract all `*.proto` and generated structures into a new sub-crate called `api`.

Most of the troubles caused by this extraction come from the various conversions for the grpc types.

e.g.

```rust
impl From<segment::types::PayloadIndexInfo> for grpc::qdrant::PayloadSchemaInfo {
    fn from(schema: segment::types::PayloadIndexInfo) -> Self {
        PayloadSchemaInfo {
            data_type: match schema.data_type {
                segment::types::PayloadSchemaType::Keyword => PayloadSchemaType::Keyword,
                segment::types::PayloadSchemaType::Integer => PayloadSchemaType::Integer,
                segment::types::PayloadSchemaType::Float => PayloadSchemaType::Float,
                segment::types::PayloadSchemaType::Geo => PayloadSchemaType::Geo,
            }
            .into(),
        }
    }
}
```

Because of Rust's coherence rule on [trait implementations](https://doc.rust-lang.org/reference/items/implementations.html#trait-implementation-coherence), the `conversions.rs` cannot be placed anymore in the root crate because the grpc types and domain types now live in different sub-crates.

The `conversions.rs` file currently in the root crate must be broken up so that each conversions `From` implementation live in the appropriate domain crate.

Given that we want to keep `segment` free of additional dependency, we will form the following dependency tree.

```mermaid
graph TD;
      segment-->storage;
      collection-->storage;
      segment-->collection;
      segment --> root;
      storage --> root;
      collection --> root;
      api --> root;
      api --> collection;
      api --> storage;      
     segment --> api;
```

Each crate will host its conversions except for `api` which will also have the ones from `segment`

A particular effort has been spent on using fully qualified names for the non-local types in `From` implementation for more clarity and to avoid naming clashes.

generall commented on build.rs:
```diff
@@ -1,9 +1,9 @@
         .compile(
-            &["src/tonic/proto/qdrant.proto"], // proto entry point
-            &["src/tonic/proto"], // specify the root location to search proto dependencies
+            &["lib/api/src/grpc/proto/qdrant.proto"], // proto entry point
+            &["lib/api/src/grpc/proto"], // specify the root location to search proto dependencies
```
Could the whole `build.rs` be moved into `api` crate instead?

generall reviewed (commented):
What do you think about splitting `*.proto` definitions into multiple parts?

From my point of view - it is inconvenient that collection APIs are defined higher in hierarchy than actual collection. 

It might be better to only keep points data structures in the `api` sub-crate, but all definitions of user-level services move back to root. In this case we will keep the source-of-truth for data structures and keep the proper hierarchy at the same time.

In further PRs we would be able to implement intercom serviced inside the `api` and they won't overlap with user-oriented services in root.

The only obstacle I can see is if it is possible to import from some arbitrary path in protobuf?

agourlay commented:
@generall I have moved the `build.rs` file into the `api` crate and I have split the proto. files within the `api` crate.

At this point the PR is still green.

Now I am trying to move the `*_service.proto` into the root crate. 

> The only obstacle I can see is if it is possible to import from some arbitrary path in protobuf?

Yes, that was a good call, I am having issues exactly with that.

```
"protoc failed: ../../lib/api/src/grpc/proto/collections.proto: Backslashes, consecutive slashes, \".\", or \"..\" are not allowed in the virtual path
```
